watchtower/wtdb/migration4/client_db.go

```diff
@@ -0,0 +1,633 @@
+// migrated per db transaction.
+const DefaultSessionsPerTx = 5000
+
+// MigrateAckedUpdates migrates the tower client DB. It takes the individual
+// Acked Updates that are stored for each session and re-stores them using the
+// RangeIndex representation.
+func MigrateAckedUpdates(sessionsPerTx int) func(kvdb.Backend) error {
+	return func(db kvdb.Backend) error {
+		log.Infof("Migrating the tower client db to move all Acked " +
+			"Updates to the new Range Index representation.")
+
+		// Migrate the old acked-updates.
+		err := migrateAckedUpdates(db, sessionsPerTx)
+		if err != nil {
+			return fmt.Errorf("migration failed: %w", err)
+		}
+
+		log.Infof("Migrating old session acked updates finished, now " +
+			"checking the migration results...")
+
+		// Before we can safety delete the old buckets, we perform a
+		// check to make sure the sessions have been migrated as
+		// expected.
+		err = kvdb.View(db, validateMigration, func() {})
+		if err != nil {
+			return fmt.Errorf("validate migration failed: %w", err)
+		}
+
+		// Delete old acked updates.
+		err = kvdb.Update(db, deleteOldAckedUpdates, func() {})
+		if err != nil {
+			return fmt.Errorf("failed to delete old acked "+
+				"updates: %w", err)
+		}
+
+		return nil
+	}
+}
+
+// migrateAckedUpdates migrates the acked updates of each session in the
+// wtclient db into the new RangeIndex form. This is done over multiple db
+// transactions in order to prevent the migration from taking up too much RAM.
+// The sessionsPerTx parameter can be used to set the maximum number of sessions
+// that should be migrated per transaction.
+func migrateAckedUpdates(db kvdb.Backend, sessionsPerTx int) error {
+	// Get migration progress stats.
+	total, migrated, err := logMigrationStats(db)
+	if err != nil {
+		return err
+	}
+	log.Infof("Total sessions=%d, migrated=%d", total, migrated)
+
+	// Exit early if the old session acked updates have already been
+	// migrated and deleted.
+	if total == 0 {
+		log.Info("Migration already finished!")
+		return nil
+	}
+
+	var (
+		finished bool
+		startKey []byte
+	)
+	for {
+		// Process the migration.
+		err = kvdb.Update(db, func(tx kvdb.RwTx) error {
+			startKey, finished, err = processMigration(
+				tx, startKey, sessionsPerTx,
+			)
+
+			return err
+		}, func() {})
+		if err != nil {
+			return err
+		}
+
+		if finished {
+			break
+		}
+
+		// Each time we finished the above process, we'd read the stats
+		// again to understand the current progress.
+		total, migrated, err = logMigrationStats(db)
+		if err != nil {
+			return err
+		}
+
+		// Calculate and log the progress if the progress is
+		// less than one.
+		//nolint:gomnd
+		progress := float64(migrated) / float64(total) * 100
+		if progress >= 100 { //nolint:gomnd
+			log.Infof("Migrating tower client DB Acked Updates " +
+				"to use the new Range Index structure " +
+				"complete, now checking the migration " +
```

Comment:
nit: this "now checking" might as well go to the validation part too.

Reply:
indeed! I see now that I actually had a log line there already. So just gonna remove this one